hydra-cluster/exe/hydra-cluster/Main.hs

```diff
@@ -24,20 +26,30 @@ run options =
     withStateDirectory $ \workDir ->
       case knownNetwork of
         Just network -> do
-          when (useMithril == UseMithril) $
-            downloadLatestSnapshotTo (contramap FromMithril tracer) network workDir
-          withCardanoNodeOnKnownNetwork fromCardanoNode workDir network $ \node -> do
-            waitForFullySynchronized fromCardanoNode node
-            publishOrReuseHydraScripts tracer node
-              >>= singlePartyHeadFullLifeCycle tracer workDir node
-        Nothing ->
+          maybeRunningNode <- findRunningCardanoNode workDir network
+          maybe
+            ( do
+                when (useMithril == UseMithril) $ do
+                  removeDirectoryRecursive $ workDir </> "db"
+                  downloadLatestSnapshotTo (contramap FromMithril tracer) network workDir
+                withCardanoNodeOnKnownNetwork fromCardanoNode workDir network $
+                  runSmokeTest tracer workDir fromCardanoNode
+            )
+            (runSmokeTest tracer workDir fromCardanoNode)
+            maybeRunningNode
+        Nothing -> do
           withCardanoNodeDevnet fromCardanoNode workDir $ \node -> do
             txId <- publishOrReuseHydraScripts tracer node
             singlePartyOpenAHead tracer workDir node txId $ \HydraClient{} -> do
               forever (threadDelay 60) -- do nothing
  where
   Options{knownNetwork, stateDirectory, publishHydraScripts, useMithril} = options
 
+  runSmokeTest tracer workDir fromCardanoNode node = do
```

Comment:
Should: avoid passing in two tracers

This is a smell of this not being a good function to factor out. Instead the finding or starting a node would be better to factor out. See my suggestion above for a better split.